File: leetcode/JavaScript/No124.binary-tree-maximum-path-sum.js
/**
 * Difficulty:
 * Hard
 *
 * Desc:
 * Given a binary tree, find the maximum path sum.
 * For this problem,
 * a path is defined as any sequence of nodes from some starting node to any node in the tree along the parent-child connections.
 * The path must contain at least one node and does not need to go through the root.
 *
 * Example 1:
 * Given the below binary tree [1,2,3],
       1
      / \
     2   3
 * Return 6.
 *
 * Example 2:
 * Given the below binary tree [-10,9,20,null,null,15,7],
      -10
      / \
     9  20
     /  \
    15   7
 * Return 42.
 *
 * 给定一个非空二叉树，返回其最大路径和。注意，是路径，即从一个节点走到另一个节点，不能走重复路线
 * 本题中，路径被定义为一条从树中任意节点出发，达到任意节点的序列。该路径至少包含一个节点，且不一定经过根节点
 */

/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var maxPathSum = function(root) {
  let result = -Infinity

  const getMaxSum = (node) => {
    if (!node) return -Infinity
    if (!node.left && !node.right) return node.val
    const maxLeft = getMaxSum(node.left)
    const maxRight = getMaxSum(node.right)

    const sum = Math.max(
      node.val,
      node.val + maxLeft,
      node.val + maxRight
    )

    result = Math.max(
      sum,
      result,
      maxLeft,
      maxRight,
      node.val + maxLeft + maxRight
    )

    return sum
  }
  const final = getMaxSum(root)
  return Math.max(result, final)
}

File: leetcode/C#/No124.binary-tree-maximum-path-sum.cs
/*
 * Difficulty:
 * Hard
 *
 * Desc:
 * Given a non-empty binary tree, find the maximum path sum.
 * For this problem, a path is defined as any sequence of nodes from some starting node to any node in the tree along the parent-child connections.
 * The path must contain at least one node and does not need to go through the root.
 *
 * Example 1:
 * Input: [1,2,3]
 *       1
 *      / \
 *     2   3
 * Output: 6
 *
 * Example 2:
 * Input: [-10,9,20,null,null,15,7]
 *   -10
 *   / \
 *  9  20
 *    /  \
 *   15   7
 * Output: 42
*/

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public int MaxPathSum(TreeNode root) {
        int res = int.MinValue;

        int MathSum(TreeNode node) {
            if (node == null) return int.MinValue;
            
            int maxLeft = MathSum(node.left);
            int maxRight = MathSum(node.right);
            int maxSum = node.val;

            try {
                maxSum = Math.Max(
                    checked(maxLeft + node.val),
                    maxSum
                );
            } catch (System.OverflowException e) {}
            try {
                maxSum = Math.Max(
                    checked(maxRight + node.val),
                    maxSum
                );
            } catch (System.OverflowException e) {}

            res = new int[]{
                res,
                maxSum,
                maxLeft,
                maxRight,
            }.Max();

            try {
                res = Math.Max(res, checked(node.val + maxLeft + maxRight));
            } catch (System.OverflowException e) {}

            return maxSum;
        }

        MathSum(root);
        return res;
    }
}